File: node-sdk/src/resources/messages.ts
import type { HttpClient } from '../http.js';
import type {
  SmartSpaceMessage,
  SendMessageParams,
  ListMessagesParams,
  SendAndWaitOptions,
  SendAndWaitResponse,
  StreamEvent,
} from '../types.js';
import { SSEStream } from '../sse.js';

export class MessagesResource {
  constructor(private http: HttpClient) {}

  async send(smartSpaceId: string, params: SendMessageParams): Promise<{
    message: SmartSpaceMessage;
    runs: Array<{ runId: string; agentEntityId: string }>;
  }> {
    return this.http.post(`/api/smart-spaces/${smartSpaceId}/messages`, params);
  }

  async list(smartSpaceId: string, params?: ListMessagesParams): Promise<{ messages: SmartSpaceMessage[] }> {
    return this.http.get(`/api/smart-spaces/${smartSpaceId}/messages`, {
      afterSeq: params?.afterSeq,
      beforeSeq: params?.beforeSeq,
      limit: params?.limit,
    });
  }

  /**
   * Send a message and wait for the agent to finish responding.
   *
   * Sends the message first (to get the runId), then subscribes to:
   * - The **run stream** for completion/failure signals
   * - The **space stream** for text deltas (from sendSpaceMessage interception)
   */
  async sendAndWait(smartSpaceId: string, params: SendAndWaitOptions): Promise<SendAndWaitResponse> {
    const timeout = params.timeout ?? 30000;
    const { timeout: _, ...sendParams } = params;

    // 1. Send the message and get the triggered runId
    const { runs } = await this.send(smartSpaceId, sendParams);

    if (!runs || runs.length === 0) {
      return { text: '', toolCalls: [] };
    }

    const runId = runs[0].runId;

    return new Promise<SendAndWaitResponse>((resolve, reject) => {
      let text = '';
      const toolCalls: SendAndWaitResponse['toolCalls'] = [];
      const activeToolCalls = new Map<string, { id: string; name: string; input: unknown; output?: unknown }>();
      let settled = false;

      // Subscribe to space stream for persisted messages (v2: no text streaming, text arrives as smartSpace.message)
      const spaceStream = new SSEStream({
        url: this.http.buildUrl(`/api/smart-spaces/${smartSpaceId}/stream`),
        headers: this.http.getAuthHeaders(),
        reconnect: false,
      });

      // Subscribe to run stream for completion signals and tool events
      const runStream = new SSEStream({
        url: this.http.buildUrl(`/api/runs/${runId}/stream`),
        headers: this.http.getAuthHeaders(),
        reconnect: false,
      });

      const timer = setTimeout(() => {
        if (!settled) {
          settled = true;
          cleanup();
          reject(new Error(`sendAndWait timed out after ${timeout}ms`));
        }
      }, timeout);

      const cleanup = () => {
        clearTimeout(timer);
        spaceStream.close();
        runStream.close();
      };

      // Space stream: collect agent text from persisted messages (v2 — no text streaming)
      spaceStream.on('smartSpace.message', (event: StreamEvent) => {
        const msg = (event.data as Record<string, unknown>)?.message as Record<string, unknown> | undefined;
        if (!msg) return;
        const role = msg.role as string;
        const agentEntityId = event.agentEntityId || (msg.entityId as string);
        // Only collect agent (assistant) messages, not the human message we just sent
        if (role !== 'assistant' && role !== 'agent') return;
        const content = msg.content as string | undefined;
        if (content) text += (text ? '\n' : '') + content;
      });

      // Run stream: tool events (v2 names)
      runStream.on('tool-call.start', (event: StreamEvent) => {
        const data = event.data as Record<string, unknown>;
        const callId = String(data.toolCallId || '');
        if (!callId) return;
        activeToolCalls.set(callId, {
          id: callId,
          name: String(data.toolName || ''),
          input: undefined,
        });
      });

      runStream.on('tool-call.complete', (event: StreamEvent) => {
        const data = event.data as Record<string, unknown>;
        const callId = String(data.toolCallId || '');
        const existing = activeToolCalls.get(callId);
        if (existing) {
          existing.input = data.input ?? existing.input;
          existing.output = data.output;
        }
      });

      // Run stream: completion signals
      runStream.on('run.completed', () => {
        if (!settled) {
          settled = true;
          cleanup();
          toolCalls.push(...activeToolCalls.values());
          resolve({ text, toolCalls, run: { id: runId } as any });
        }
      });

      runStream.on('run.failed', (event: StreamEvent) => {
        if (!settled) {
          settled = true;
          cleanup();
          const data = event.data as Record<string, unknown>;
          reject(new Error(String(data.errorMessage || 'Run failed')));
        }
      });

      // Error on either stream
      const onError = (event: StreamEvent) => {
        if (!settled) {
          settled = true;
          cleanup();
          reject(new Error(String((event.data as Record<string, unknown>).error || 'Stream error')));
        }
      };

      spaceStream.on('error', onError);
      runStream.on('error', onError);
    });
  }
}

File: hsafa-core/old-hsafa/python-sdk/hsafa/resources/messages.py
import threading
from typing import Any, Dict, List, Optional

from ..http import HttpClient
from ..sse import SSEStream


class MessagesResource:
    def __init__(self, http: HttpClient):
        self._http = http

    def send(
        self,
        space_id: str,
        content: str,
        entity_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        role: Optional[str] = None,
        trigger_agents: Optional[bool] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"content": content}
        if entity_id is not None:
            body["entityId"] = entity_id
        if metadata is not None:
            body["metadata"] = metadata
        if role is not None:
            body["role"] = role
        if trigger_agents is not None:
            body["triggerAgents"] = trigger_agents
        return self._http.post(f"/api/smart-spaces/{space_id}/messages", body)

    def list(
        self,
        space_id: str,
        after_seq: Optional[str] = None,
        before_seq: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> Dict[str, Any]:
        return self._http.get(
            f"/api/smart-spaces/{space_id}/messages",
            {"afterSeq": after_seq, "beforeSeq": before_seq, "limit": limit},
        )

    def send_and_wait(
        self,
        space_id: str,
        content: str,
        entity_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: float = 30.0,
    ) -> Dict[str, Any]:
        """
        Send a message and block until the agent finishes responding.
        Returns { 'text': str, 'tool_calls': list, 'run': dict }.
        Raises TimeoutError if the agent doesn't respond within `timeout` seconds.
        """
        result = self.send(space_id, content, entity_id=entity_id, metadata=metadata)
        runs: List[Dict[str, Any]] = result.get("runs", [])

        if not runs:
            return {"text": "", "tool_calls": [], "run": None}

        run_id: str = runs[0]["runId"]

        text_parts: List[str] = []
        tool_calls: Dict[str, Dict[str, Any]] = {}
        done_event = threading.Event()
        error_holder: Dict[str, Any] = {}
        settled = threading.Lock()
        _settled = [False]

        def finish() -> None:
            with settled:
                if not _settled[0]:
                    _settled[0] = True
                    done_event.set()

        def on_error_event(event: Dict[str, Any]) -> None:
            with settled:
                if not _settled[0]:
                    _settled[0] = True
                    error_holder["error"] = str(
                        (event.get("data") or {}).get("error", "Stream error")
                    )
                    done_event.set()

        space_stream = SSEStream(
            self._http.build_url(f"/api/smart-spaces/{space_id}/stream"),
            self._http.get_auth_headers(),
            reconnect=False,
        )
        run_stream = SSEStream(
            self._http.build_url(f"/api/runs/{run_id}/stream"),
            self._http.get_auth_headers(),
            reconnect=False,
        )

        def on_space_message(event: Dict[str, Any]) -> None:
            msg = (event.get("data") or {}).get("message") or {}
            role_ = msg.get("role", "")
            if role_ not in ("assistant", "agent"):
                return
            c = msg.get("content")
            if c:
                text_parts.append(c)

        def on_agent_inactive(event: Dict[str, Any]) -> None:
            rid = event.get("runId") or (event.get("data") or {}).get("runId")
            if rid == run_id:
                finish()

        def on_tool_started(event: Dict[str, Any]) -> None:
            data = event.get("data") or {}
            call_id = str(data.get("streamId") or data.get("toolCallId") or "")
            if call_id:
                tool_calls[call_id] = {
                    "id": call_id,
                    "name": str(data.get("toolName", "")),
                    "input": None,
                    "output": None,
                }

        def on_tool_done(event: Dict[str, Any]) -> None:
            data = event.get("data") or {}
            call_id = str(data.get("streamId") or data.get("toolCallId") or "")
            if call_id in tool_calls:
                tool_calls[call_id]["input"] = data.get("input")
                tool_calls[call_id]["output"] = data.get("result") or data.get("output")

        def on_run_completed(_event: Dict[str, Any]) -> None:
            finish()

        def on_run_failed(event: Dict[str, Any]) -> None:
            with settled:
                if not _settled[0]:
                    _settled[0] = True
                    data = event.get("data") or {}
                    error_holder["error"] = str(data.get("errorMessage", "Run failed"))
                    done_event.set()

        space_stream.on("space.message", on_space_message)
        space_stream.on("agent.inactive", on_agent_inactive)
        space_stream.on("error", on_error_event)

        run_stream.on("tool.started", on_tool_started)
        run_stream.on("tool.done", on_tool_done)
        run_stream.on("run.completed", on_run_completed)
        run_stream.on("run.failed", on_run_failed)
        run_stream.on("error", on_error_event)

        space_stream.start()
        run_stream.start()

        try:
            if not done_event.wait(timeout=timeout):
                raise TimeoutError(f"send_and_wait timed out after {timeout}s")
            if "error" in error_holder:
                raise RuntimeError(error_holder["error"])
            return {
                "text": "\n".join(text_parts),
                "tool_calls": list(tool_calls.values()),
                "run": {"id": run_id},
            }
        finally:
            space_stream.close()
            run_stream.close()
